### DISCUSSION TITLE
Spinning up NorESM3

### DISCUSSION THREAD

**Discussion by mvertens** [NorESMhub/noresm3_dev_simulations#58] [General]
The spinup process for NorESM involves running a fully coupled run and setting mediator namelist to create auxiliary history output that can then be ingested by data components to force the target prognostic components.

-----

### **1. Creating mediator history files**

- **create_newcase command for fully coupled but with DGLC**
```
 ./create_newcase --case <case> --compset 1850_CAM%DEV%LT%NORESM%CAMoslo_CLM60%FATES-SP_CICE_BLOM%ECO_MOSART_DGLC%NOEVOLVE_SWAV_SESP  \\
 --res ne30pg3_tn14 \\
 --project <project> --run-unsupported --mach betzy 
```

- **user_nl_cpl**
```
 histaux_atm2med_file1_enabled = .true.
 histaux_atm2med_file2_enabled = .true.
 histaux_atm2med_file3_enabled = .true.
 histaux_atm2med_file4_enabled = .true.
 histaux_atm2med_file5_enabled = .true.
 histaux_rof2med_file1_enabled = .true.
 histaux_atm2med_file5_ntperfile = 1
 histaux_atm2med_file5_history_n = 1
 histaux_atm2med_file5_history_option = 'ndays'
 histaux_rof2med_file1_ntperfile = 1
 histaux_rof2med_file1_history_n = 1
 histaux_rof2med_file1_history_option = 'ndays'
 histaux_l2x1yrg = .true.
```

Using the generated mediatory history files as input to DATM for forcing CLM, input to DATM and DROF for forcing BLOM and input to DLND for forcing CISM.

The output from the mediator history files is as follows:

- **DATM output to spinup CLM and BLOM -** that contains the following variables in the input file nuopc.runconfig
  - **instantaneous fields** 
    ```
     histaux_atm2med_file1_auxname = atm.1h.inst
     histaux_atm2med_file1_doavg = .true.
     histaux_atm2med_file1_enabled = .false.
     histaux_atm2med_file1_flds = Faxa_swndr:Faxa_swvdr:Faxa_swndf:Faxa_swvdf
     histaux_atm2med_file1_history_n = 1
     histaux_atm2med_file1_history_option = nhours
     histaux_atm2med_file1_ntperfile = 24
          result in files
     <case>cpl.hx.atm.1h.inst.0001-01-01-03600.nc
     <case>cpl.hx.atm.1h.inst.0001-01-02-03600.nc
     ..... 
     ```
  - **hourly averaged fields**
    ```
     histaux_atm2med_file2_auxname = atm.1h.avrg
     histaux_atm2med_file2_doavg = .true.
     histaux_atm2med_file2_enabled = .false.
     histaux_atm2med_file2_flds = Sa_u:Sa_v
     histaux_atm2med_file2_history_n = 1
     histaux_atm2med_file2_history_option = nhours
     histaux_atm2med_file2_ntperfile = 24
          results in the files
     <case>.cpl.hx.atm.1h.avrg.0001-01-01-03600.nc
     <case>.cpl.hx.atm.1h.avrg.0001-01-02-03600.nc
      ....
    ```
  - **3-hourly averaged fields precipitation fields**
     ```
     histaux_atm2med_file3_auxname = atm.3hprec.avrg
     histaux_atm2med_file3_doavg = .true.
     histaux_atm2med_file3_enabled = .true.
     histaux_atm2med_file3_flds = Faxa_rainc:Faxa_rainl:Faxa_snowc:Faxa_snowl
     histaux_atm2med_file3_history_n = 3
     histaux_atm2med_file3_history_option = nhours
     histaux_atm2med_file3_ntperfile = 8
         results in files
     <case>.cpl.hx.atm.3hprec.avrg.0001-01-01-10800.nc
     <case>.cpl.hx.atm.3hprec.avrg.0001-01-02-10800.nc
      ...
    ```

   - **3-hourly averaged fields other than precipitation**
     ```
     histaux_atm2med_file4_auxname = atm.3h.avrg
     histaux_atm2med_file4_doavg = .true.
     histaux_atm2med_file4_enabled = .true.
     histaux_atm2med_file4_flds = Sa_z:Sa_topo:Sa_u:Sa_v:Sa_tbot:Sa_ptem:Sa_shum:Sa_dens:Sa_pbot:Sa_pslv:Faxa_lwdn:Faxa_rainc:Faxa_rainl:Faxa_snowc:Faxa_snowl:Faxa_swndr:Faxa_swvdr:Faxa_swndf:Faxa_swvdf:Sa_co2diag:Sa_co2prog
     histaux_atm2med_file4_history_n = 3
     histaux_atm2med_file4_history_option = nhours
     histaux_atm2med_file4_ntperfile = 8
          results in files 
     cpl.hx.atm.3h.avrg.0001-01-01-10800.nc
     cpl.hx.atm.3h.avrg.0001-01-02-10800.nc
     .....
     ```

  - **daily averaged fields** 
     ```
     histaux_atm2med_file5_auxname = atm.24h.avrg
     histaux_atm2med_file5_doavg = .true.
     histaux_atm2med_file5_enabled = .false.
     histaux_atm2med_file5_flds = Faxa_bcph:Faxa_ocph:Faxa_dstwet:Faxa_dstdry:Sa_co2prog:Sa_co2diag
     histaux_atm2med_file5_history_n = 3
     histaux_atm2med_file5_history_option = nhours
     histaux_atm2med_file5_ntperfile = 2
            results in files
    <case>.cpl.hx.atm.24h.avrg.0001-01-02-00000.nc
    <case>.cpl.hx.atm.24h.avrg.0001-01-03-00000.nc
    <case>.cpl.hx.atm.24h.avrg.0003-01-01-00000.nc
    ```

- **ROF output to spinup BLOM (daily averaged fields ) -** that contains the following variables in the input file nuopc.runconfig
     ```
     histaux_rof2med_file1_auxname = rof.24h.avrg
     histaux_rof2med_file1_doavg = .true.
     histaux_rof2med_file1_enabled = .true.
     histaux_rof2med_file1_flds = all
     histaux_rof2med_file1_history_n = 1
     histaux_rof2med_file1_history_option = ndays
     histaux_rof2med_file1_ntperfile = 1
        results in files 
     <case>.cpl.hx.rof.24h.avrg.0002-01-01-00000.nc
     <case>.cpl.hx.rof.24h.avrg.0003-01-01-00000.nc
     .....
     ```

-----

### **2. Spinning up CTSM using above auxiliary files**
NOTE: this assumes that noresm2_5_alpha09 will be used - since changes to cdeps had to be made to have this working
Assumptions for this example:
```
1. spinup data is in directory /cluster/work/users/mvertens/noresm/spinup_tests/spinup_data
2. the casename for the spinup data is n1850.ne30_tn14.hybrid_fatessp_cplhist.202401031
```
```
> ./create_newcase --case <casepath> --compset 1850_DATM%CPLHIST_CLM60%BGC-CROP_SICE_SOCN_SROF_SGLC_SWAV --res ne30pg3_ne30pg3_mtn14 --project nn9039k --run-unsupported --mach betzy

> cd <casepath>

> ./xmlchange NTASKS=512
> ./xmlchange NTASKS_OCN=354
> ./xmlchange DATM_CPLHIST_CASE=n1850.ne30_tn14.hybrid_fatessp_cplhist.202401031
> ./xmlchange DATM_CPLHIST_DIR=/cluster/work/users/mvertens/noresm/spinup_tests/spinup_data
> ./xmlchange DATM_YR_START=1
> ./xmlchange DATM_YR_END=1
> ./xmlchange DATM_PRESNDEP=none

> ./case.setup
> ./caset.build
> ./case.submit
```
-----

### **3. Spinning up BLOM using above auxiliary files**
NOTE: this assumes that noresm2_5_alpha09 will be used - since changes to cdeps had to be made to have this working
Assumptions for this example:
```
1. spinup data is in dirctory /cluster/work/users/mvertens/noresm/spinup_tests/spinup_data
2. the casename for the spinup data is n1850.ne30_tn14.hybrid_fatessp_cplhist.202401031
```
 
```
>  ./create_newcase --case <casepath> --compset 1850_DATM%CPLHIST_SLND_CICE_BLOM%ECO_DROF%CPLHIST_SGLC_SWAV --res ne30pg3_tn14 --project <project> --run-unsupported --mach betzy

> cd <casepath>

> ./xmlchange NTASKS=512
> ./xmlchange NTASKS_OCN=354
> ./xmlchange DATM_CPLHIST_CASE=n1850.ne30_tn14.hybrid_fatessp_cplhist.202401031
> ./xmlchange DATM_CPLHIST_DIR=/cluster/work/users/mvertens/noresm/spinup_tests/spinup_data
> ./xmlchange DATM_YR_START=1
> ./xmlchange DATM_YR_END=1
> ./xmlchange DATM_PRESNDEP=clim_1850
> ./xmlchange DROF_CPLHIST_CASE=n1850.ne30_tn14.hybrid_fatessp_cplhist.202401031
> ./xmlchange DROF_CPLHIST_DIR=/cluster/work/users/mvertens/noresm/spinup_tests/spinup_data

> ./case.setup
> ./case.build
> ./case.submit

```
-----

### **4. Spinning up CISM using above auxiliary files**
NOTE: this assumes that noresm2_5_alpha09 will be used - since changes to cdeps had to be made to have this working
Assumptions for this example:
```
1. spinup data is in dirctory /cluster/work/users/mvertens/noresm/spinup_tests/spinup_data
2. the casename for the spinup data is n1850.ne30_tn14.hybrid_fatessp_cplhist.202401031
```
NOTE: changes must still be implemented in CDEPS/CMEPS in order to have docn multiple levels be written to auxiliary stream files

-----
### **5. Re-coupling after BLOM spinup**
This procedure has been tested in issue #297 

1.  Make sure that the location for the BLOM restart file also includes the HAMOCC restart file, e.g.

> ICFILE = "/cluster/shared/noresm/inputdata/restart/n1850.ne16_tn14.BLOM_cplhist_hybrid.noresm3_0_beta06.20251117/n1850.ne16_tn14.BLOM_cplhist_hybrid.noresm3_0_beta06.20251117.blom.r.0751-01-01-00000.nc"
> /cluster/shared/noresm/inputdata/restart/n1850.ne16_tn14.BLOM_cplhist_hybrid.noresm3_0_beta06.20251117/n1850.ne16_tn14.BLOM_cplhist_hybrid.noresm3_0_beta06.20251117.blom.rbgc.0751-01-01-00000.nc

2. Specify consistent input file for CICE, e.g.

> ice_ic = '/nird/datapeak/NS9560K/noresm3/cases/n1850.ne16_tn14.BLOM_cplhist_hybrid.noresm3_0_beta06.20251117/rest/0771-01-01-00000/n1850.ne16_tn14.BLOM_cplhist_hybrid.noresm3_0_beta06.20251117.cice.r.0771-01-01-00000.nc'

3. Disable sediment spinup, and set BLOM/iHAMOCC output size to "spinup" to reduce the amount of otput produced by the ocean module.

> ./xmlchange HAMOCC_SEDSPINUP=FALSE
> ./xmlchange BLOM_OUTPUT_SIZE=spinup
> ./xmlchange HAMOCC_OUTPUT_SIZE=spinup

**Comment by adagj** [comment]
@mvertens I suggest to move this recipe to the wiki

**Comment by mvertens** [comment]
Right. I agree. My plan is to add the instructions for the second part here and then resolve any questions then move it to the wiki once everything was clear. Does that sound reasonable?

**Comment by adagj** [comment]
@TomasTorsvik @tjiputra 

This discussion is probably useful for the spin up

**Comment by TomasTorsvik** [comment]
I'm planning to test this procedure after merging `noresm3_0_alpha03`

**Comment by mvertens** [comment]
I'm happy to provide help debugging this if I am feeling okay. I've been taking to @TomasTorsvik about this last week. I think once all of this is validated - then we should move the contents of this discussion to the Wiki.

**Comment by TomasTorsvik** [comment]
@mvertens , @gold2718 , @mvdebolskiy , @matsbn , @tjiputra , @JorgSchwinger , @mpetrini-norce , @hgoelzer 

Hi,
I have tested the procedure outlined above, and have completed stages 1 (#109) and 2 (#119). However, I have not been able to complete stage 3 (BLOM), and I seem to be missing some information for stage 4 (CISM).

I did the stage 1 run for 8 years, and have tested spinning up CTSM with both 1 year and 8 year repeat cycles. The DATM settings for the CTSM run are as follows:

**DATM settings for spinning up CTSM**
```
$ ./xmlquery -p DATM

Results in group run_component_datm
        DATM_CO2_TSERIES: none
        DATM_CPLHIST_CASE: n1850.ne30_tn14.hybrid_fates-nocomp_spinup.3.0a02.20250415
        DATM_CPLHIST_DIR: /cluster/shared/noresm/inputdata/cplhist/noresm3_0/n1850.ne30_tn14.hybrid_fates-nocomp_spinup.3.0a02.20250415
        DATM_CPLHIST_DOMAIN_FILE: null
        DATM_MODE: CPLHIST
        DATM_PRESAERO: cplhist
        DATM_PRESNDEP: none
        DATM_PRESO3: clim_1850
        DATM_SKIP_RESTART_READ: FALSE
        DATM_TOPO: cplhist
        DATM_YR_ALIGN: 1
        DATM_YR_END: 1
        DATM_YR_START: 1
        DATM_YR_START_FILENAME: 9999
```

When I try spinning up BLOM, I get a error that I have traced to `CDEPS/streams/dshr_stream_mod.F90` (LVD = "last valid date")
```
(shr_stream_findBounds) ERROR: LVD not found, all data is after yearLast
```
I found that I needed to put in some values for `DROF_CPLHIST_YR_START` and `DROF_CPLHIST_YR_END`. I have also tried setting `DATM_YR_ALIGN` and `DROF_CPLHIST_YR_ALIGN` to 1850, but this does not solve the problem. I haven't looked into the CDEPS code yet, I thought to ask around first in case someone already has a quick fix for this issue.

I also noted that there is no variable `DROF_CPLHIST_DIR`. It seems the model use `DATM_CPLHIST_DIR` for both DATM/DROF. If there is supposed to be a `DROF_CPLHIST_DIR` variable, could there be an issue with the CDEPS version?

**DATM settings for spinning up BLOM**
```
$ ./xmlquery -p DATM

Results in group run_component_datm
        DATM_CO2_TSERIES: none
        DATM_CPLHIST_CASE: n1850.ne30_tn14.hybrid_fates-nocomp_spinup.3.0a02.20250415
        DATM_CPLHIST_DIR: /cluster/shared/noresm/inputdata/cplhist/noresm3_0/n1850.ne30_tn14.hybrid_fates-nocomp_spinup.3.0a02.20250415
        DATM_CPLHIST_DOMAIN_FILE: null
        DATM_MODE: CPLHIST
        DATM_PRESAERO: cplhist
        DATM_PRESNDEP: clim_1850
        DATM_PRESO3: none
        DATM_SKIP_RESTART_READ: FALSE
        DATM_TOPO: cplhist
        DATM_YR_ALIGN: 1
        DATM_YR_END: 1
        DATM_YR_START: 1
        DATM_YR_START_FILENAME: 9999
```

**DROF settings for spinning up BLOM**
```
$ ./xmlquery -p DROF

Results in group run_component_drof
        DROF_CPLHIST_CASE: n1850.ne30_tn14.hybrid_fates-nocomp_spinup.3.0a02.20250415
        DROF_CPLHIST_YR_ALIGN: 1
        DROF_CPLHIST_YR_END: 1
        DROF_CPLHIST_YR_START: 1
        DROF_MODE: CPLHIST
        DROF_SKIP_RESTART_READ: FALSE
```

Finally, what compset specifications should I use for the CISM spinup?

**Reply by hgoelzer** [reply]
Thanks for looking into this. For CISM, I think we (Michele and/or me) will have to go through the process ourselves at least once to document this properly for NorESM3. The process has changed since NorESM2 (where we did this partially offline). We will be able to look into that next week (back from EGU and holidays)

**Comment by TomasTorsvik** [comment]
@mvertens , @gold2718 , @mvdebolskiy , @matsbn , @tjiputra , @JorgSchwinger
I checked again `Buildconf` files. The `datm.iup_data_list` includes a long list of files from `*.cpl.hx.atm.*`, whereas `drof.input_data_list` contains
```
$ cat Buildconf/drof.input_data_list 
model_meshfile = /cluster/shared/noresm/inputdata/share/meshes/r05_nomask_c110308_ESMFmesh.nc
model_maskfile = /cluster/shared/noresm/inputdata/share/meshes/r05_nomask_c110308_ESMFmesh.nc
mesh = /cluster/shared/noresm/inputdata/share/meshes/r05_nomask_c110308_ESMFmesh.nc
file1 = /cluster/shared/noresm/inputdata/lnd/dlnd7/RX1/runoff.daitren.iaf.20120419.nc
```
I would expect a list of `file1` ... `fileN` with reference to `*.cpl.hx.rof.*` instead of the reference to `runoff.daitren.iaf.20120419.nc`.

**Comment by mvertens** [comment]
@TomasTorsvik - can you please point me to your case directory and your run directory.

**Comment by TomasTorsvik** [comment]
@mvertens - I think CDEPS was not up to date in `noresm3_0_alpha02`. I'm trying again now with the latest tag `noresm3_0_aplha03b`.

Old case:
- case directory: `/cluster/projects/nn9560k/tomast/NorESM/cases/n1850.ne30_tn14.BLOM_cplhist.3.0a02.20250508`
- run directory: `/cluster/work/users/tomast/noresm/n1850.ne30_tn14.BLOM_cplhist.3.0a02.20250508`

New case:
- case directory: `/cluster/projects/nn9560k/tomast/NorESM/cases/n1850.ne30_tn14.BLOM_cplhist.3.0a03b.20250513`
- run directory: `/cluster/work/users/tomast/noresm/n1850.ne30_tn14.BLOM_cplhist.3.0a03b.20250513`

The new case is picking up reference to `*.cpl.hx.rof.*` input files.

**Reply by mvertens** [reply]
@TomasTorsvik - thanks. In checking your run directory - it looks like the case is running.

**Comment by TomasTorsvik** [comment]
@mvertens , @tjiputra - Do you have a suggestion how to plug in again the output from CTSM and BLOM/CICE spinup? Is it suficient to just change the respective `ICFILE` settings for each component?

**Comment by mvertens** [comment]
@TomasTorsvik - I'm not sure what you are asking. Can you clarify what you mean by plugging in again the output. My understanding is that you have creating the forcing files and now you are using them to spin up CTSM and BLOM/CICE separately? What exactly do you want to change?

**Comment by TomasTorsvik** [comment]
@mvertens - I was thinking, after we have spin up model output for CTSM and BLOM/CICE. 

I assume what we would do is something like this:
* create a new case for a `n1850.ne30_tn14.*` simulation
* Import initial conditions `user_nl` settings:
```
user_nl_blom:
-------------
ICFILE = "<path to BLOM/CICE spinup>/<casename>.blom.r.<year-month-day-ts>.nc"

user_nl_cice:
-------------
ice_ic = "<path to BLOM/CICE spinup>/<casename>.cice.r.<year-month-day-ts>.nc"

user_nl_clm:
-------------
use_fates_nocomp = .true.
use_fates_fixed_biogeog = .true.
use_fates_luh = .false.
fates_radiation_model = 'twostream'
finidat = "<path to CTSM spinup>/<casename>.clm2.r.<year-month-day-ts>.nc"
fates_history_dimlevel = 1,1
```

Is this the correct procedure?

**Reply by mvertens** [reply]
@TomasTorsvik - yes that makes sense. For CAM I think you will need to point to an initial file rather than a restart file to start up the run. @gold2718 can confirm.

**Comment by gold2718** [comment]
CAM will use an initial file created during the same spinup runs. For production runs, will we include all of these 'spun-up' initial data files in a use case?

**Reply by TomasTorsvik** [reply]
@gold2718 - I suppose we will make an attempt to start production runs from these spun-up initial data files, but I guess it will depend on how large the drift is. For the ocean interior, we probably need to be somewhere close to the WOA18 reanalysis data at the start of the production runs.

**Comment by TomasTorsvik** [comment]
@JorgSchwinger , @tjiputra , @mvertens , @kjetilaas , @DirkOlivie , @oyvindseland , @adagj , @matsbn , @mvdebolskiy 
Until now, I have been running `ne30pg3_tn14` with initial conditions from repeat-year-forcing (JRA) for BLOM and CICE, and CLM spinup files that @kjetilaas has provided. I would like to get this closer to a spun-up state, and I wonder if I could use the same ocean spin-up as for `ne16pg3_tn14` as a starting point?
I suppose we will not have time for a completely separate spin-up of the ocean for `ne30pg3_tn14`, but perhaps a shorter spin-up cycle could be carried out?
